shell/platform/android/vsync_waiter_android.cc

```diff
@@ -66,8 +90,41 @@ void VsyncWaiterAndroid::OnVsyncFromNDK(int64_t frame_nanos, void* data) {
                    "frame_target_time",
                    target_time.ToEpochDelta().ToMicroseconds());
 
-  auto* weak_this = reinterpret_cast<std::weak_ptr<VsyncWaiter>*>(data);
-  ConsumePendingCallback(weak_this, frame_time, target_time);
+  auto* weak_this = reinterpret_cast<std::weak_ptr<VsyncWaiterAndroid>*>(data);
+  ConsumePendingCallback(weak_this, frame_time, target_time, target_time, 0);
+}
+
+// static
+void VsyncWaiterAndroid::OnVsyncFromNDK33(
+    AndroidChoreographer::AChoreographerFrameCallbackData* callback_data,
+    void* data) {
+  auto frame_time = AndroidChoreographer::GetFrameTime(callback_data);
+  auto now = fml::TimePoint::Now();
+  if (frame_time > now) {
+    frame_time = now;
+  }
+  auto* weak_this = reinterpret_cast<std::weak_ptr<VsyncWaiterAndroid>*>(data);
+  auto shared_this = weak_this->lock();
+  if (!shared_this) {
+    delete weak_this;
+    return;
+  }
+  // Target to present with +1 frame latency to avoid jitter from
+  // uneven frame pacing.
+  auto target_time = AndroidChoreographer::GetFrameDeadline(callback_data, 1);
+  auto next_frame_time = frame_time + fml::TimeDelta::FromNanoseconds(
+                                          1000000000.0 / g_refresh_rate_);
```

Comment:
Please cache the nanoseconds value to avoid doing the divide every frame.

Reply:
Done